---
title: "Methods and packages being tested"
author: "Anand Srinivasan"
date: "19/04/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r sparrow}
#BiocManager::install(c("sparrow", "DEFormats"))
library(sparrow)
library(DEFormats)

res <- paste0(results_dir, "deseq2_results_annotated.tsv")
results_annotated <- read.delim(res)
gdb <- sparrow::getMSigGeneSetDb(c('H', 'C2'), species = 'human', id.type = c("ensembl", "entrez", "symbol"))


genes <- dplyr::select(results_annotated, ensembl_gene_id, ENTREZID, hgnc_symbol)

design <- model.matrix(~DeSeq2_return$dds[[deseq2_design_condition]])
dge <- DGEList(counts = assay(DeSeq2_return$dds), 
               samples = data.frame(DeSeq2_return$dds$SampleName, DeSeq2_return$dds[[deseq2_design_condition]]), 
               norm.factors = NULL, 
               group = DeSeq2_return$dds[[deseq2_design_condition]], 
               genes = genes)
vm <- limma::voom(dge, design)


mg <- seas(
  gsd = gdb,
  x = vm,
  design = vm$design,
  #contrast = DeSeq2_return$dds$Treatment,
  methods = c("fgsea", "camera", "ora", "fry", "roast")
)


mg_res.fg <- sparrow::results(mg, "fgsea") %>% 
  dplyr::filter(padj < 0.05, collection == "H") %>% 
  arrange(desc(mean.logFC)) 

###%>% select(name, n, mean.logFC, padj) 

View(mg_res.fg)

iplot(mg, 'HALLMARK_INTERFERON_ALPHA_RESPONSE',
      type = 'density', value = 'logFC')
```


```{r multiGSEA}
library(multiGSEA)
#library(tidyverse)
#library( magrittr)
res <- paste0(results_dir, "deseq2_results_annotated.tsv")
results_annotated <- read.delim(res)
omics_data <- initOmicsDataStructure( layer = c("transcriptome"))
                                                
                                                                                                                                 #"proteome",
                                                #"metabolome"))
##create data frame of DEG results
transcriptome <- results_annotated %>% drop_na()

#load the layer
omics_data$transcriptome <- rankFeatures( transcriptome$log2FoldChange, 
                                          transcriptome$padj)

#
names(omics_data$transcriptome) <- transcriptome$hgnc_symbol

#get the pathways
databases <- c( "reactome")
layers <- names(omics_data)


new_pathways <- getMultiOmicsFeatures(
  #dbs = databases,
  layer = layers,
  returnTranscriptome = "SYMBOL",
  useLocal = FALSE
)

DeSeq2_return$vsd$sizeFactor
##run the gsea
enrichment_scores <- multiGSEA(new_pathways, omics_data)

#get pvalues

df <- extractPvalues( enrichmentScores = enrichment_scores,
                      pathwayNames = names(new_pathways[[1]]))

df <- cbind( data.frame( pathway = names( new_pathways[[1]])), df)

df <- df %>% dplyr::filter(transcriptome_padj < 0.05)


##enrichment scores

mGSEA_res <- as.data.frame(enrichment_scores) %>% 
  dplyr::select(transcriptome.pathway, transcriptome.NES, transcriptome.padj, transcriptome.size, transcriptome.leadingEdge) %>% 
  #dplyr::filter(transcriptome.padj < 0.05) %>% 
  arrange(transcriptome.NES)





```


```{r co_Reg_net}
if (!require("CoRegNet")) {BiocManager::install("CoRegNet")}
library(CoRegNet)

#GWENA::filter_low_var(cemitool_input %>% t(), 0.4, )

data(CIT_BLCA_EXP,HumanTF,CIT_BLCA_Subgroup)
options("mc.cores"=12)
grn_2 = hLICORN(cemitool_input, TFlist=HumanTF)

invisible(gc())
options("mc.cores"=12)
grn_2 = hLICORN(BioNERO::filter_by_variance(cemitool_input, percentile = 0.4), TFlist=HumanTF, verbose = TRUE)

BioNERO::filter_by_variance(cemitool_input, percentile = 0.4) -> a
```



```{r tigress}
library(tigress)
gc()
genenames <- rownames(ecoli$exp)
tfindices <- sort(unique(ecoli$reg[,1]))
tfnames <- genenames[tfindices]
ntf <- length(tfindices)
ntf
targetindices <- sort(unique(ecoli$reg[,2]))
targetnames <- genenames[targetindices]
ntarget <- length(targetindices)
ntarget

nstepsLARS = 20
edgepred <- tigress(t(ecoli$exp), tflist=tfnames, targetlist=targetnames, nstepsLARS = nstepsLARS)


edgepred[[2]] %>% View()
```


```{r modifier}
#BiocManager::install("MODA")
librarian::shelf("MODA", "STRINGdb")
#devtools::install_git(url = "https://gitlab.com/Gustafsson-lab/MODifieR.git", git="external")
if (!require("MODifieR")) {devtools::install_git(url = "https://gitlab.com/Gustafsson-lab/MODifieR.git", git="external")}
library("MODifieR")

MODifieR_input <- assay(DeSeq2_return$dds) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   dplyr::mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"),
      .before=1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>% 
  column_to_rownames("hgnc_symbol") %>%
  as.matrix()

MODifieR_input_rna <- create_input_rnaseq(count_matrix = MODifieR_input, 
                                          group1_indici = 1:3, 
                                          group2_indici = 4:6, 
                                          group1_label = Control, 
                                          group2_label = Test,
                                          use_adjusted = T, 
                                          normalize_quantiles = TRUE)
ppi_network2 <- ppi_network %>% 
  dplyr::mutate(
      entrez1 = mapIds(
        EnsDb.Hsapiens.v86,
        keys = entrez1,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "ENTREZID",
        column = "SYMBOL"),
      .before=1
    ) %>% #drop_na() %>%
  dplyr::mutate(
      entrez2 = mapIds(
        EnsDb.Hsapiens.v86,
        keys = entrez2,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "ENTREZID",
        column = "SYMBOL"),
      .before=1
    ) %>% drop_na()

gc()
wgnca_module <- wgcna(MODifieR_input = MODifieR_input_rna,
                      group_of_interest = 2)

wgnca_module2 <- wgcna_adjust_significance(pval_cutoff = 0.1, 
                                          wgcna_module = wgnca_module,
                                          use_unadjusted = FALSE)


wgnca_module3 <- wgcna_split_module_by_color(wgcna_module = wgnca_module2)


gc()
diffcoex_module <- diffcoex(MODifieR_input = MODifieR_input_rna)
gc()

moda_module <- moda(MODifieR_input = MODifieR_input_rna, 
                    group_of_interest = 2)

gc()
correlation_clique_module <- correlation_clique(MODifieR_input = MODifieR_input_rna, 
                                                #ppi_network = ppi_network, 
                                                iteration = 2)

ppi_network3 <- dorothea_df %>% select(source, target, weight) %>% distinct()


ppi_network2$entrez1 %>% unique() %>% length()

	

```


```{r csdR}
library(csdR)
csdR_input <- assay(DeSeq2_return$dds) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   dplyr::mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"),
      .before=1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>% 
  column_to_rownames("hgnc_symbol") %>%
  as.matrix()




csd_results <- run_csd(
    x_1 = csdR_input[, 4:6] %>% t() , x_2 = csdR_input[, 1:3] %>% t() ,
    n_it = 10, nThreads = 2L, verbose = TRUE
)
pairs_to_pick <- 100L
c_filter <- partial_argsort(csd_results$cVal, pairs_to_pick)
c_frame <- csd_results[c_filter, ]
s_filter <- partial_argsort(csd_results$sVal, pairs_to_pick)
s_frame <- csd_results[s_filter, ]
d_filter <- partial_argsort(csd_results$dVal, pairs_to_pick)
d_frame <- csd_results[d_filter, ]

csd_filter <- c_filter %>%
    union(s_filter) %>%
    union(d_filter)
csd_frame <- csd_results[csd_filter, ]


c_network <- graph_from_data_frame(c_frame, directed = FALSE)
s_network <- graph_from_data_frame(s_frame, directed = FALSE)
d_network <- graph_from_data_frame(d_frame, directed = FALSE)
E(c_network)$edge_type <- "C"
E(s_network)$edge_type <- "S"
E(d_network)$edge_type <- "D"
combined_network <- igraph::union(c_network, s_network, d_network)
# Auxillary function for combining
# the attributes of the three networks in a proper way
join_attributes <- function(graph, attribute) {
    ifelse(
        test = is.na(edge_attr(graph, glue("{attribute}_1"))),
        yes = ifelse(
            test = is.na(edge_attr(graph, glue("{attribute}_2"))),
            yes = edge_attr(graph, glue("{attribute}_3")),
            no = edge_attr(graph, glue("{attribute}_2"))
        ),
        no = edge_attr(graph, glue("{attribute}_1"))
    )
}

library(glue)
E(combined_network)$edge_type <- join_attributes(combined_network, "edge_type")
layout <- layout_nicely(combined_network)
E(combined_network)$color <- recode(E(combined_network)$edge_type,
    C = "darkblue", S = "green", D = "darkred"
)
plot(combined_network, layout = layout,
    vertex.size = 3, edge.width = 2, vertex.label.cex = 0.001)

gc()




```



```{r dcanr}
#BiocManager::install("dcanr")
library(dcanr)
#load data
data(sim102)
#get available conditions
getConditionNames(sim102)
simdata <- getSimData(sim102, cond.name = 'UME6', full = FALSE)
emat <- simdata$emat
ume6_kd <- simdata$condition


dcanr_input <- assay(DeSeq2_return$dds) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   dplyr::mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"),
      .before=1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>% 
  column_to_rownames("hgnc_symbol") %>%
  as.matrix()


colData <-  SummarizedExperiment::colData(DeSeq2_return$vsd) %>%
  as.data.frame() %>% 
  dplyr::select(deseq2_design_condition) %>%
  rownames_to_column("SampleName")

dcanr_input <- SummarizedExperiment(
  assays = dcanr_input,
  rowData = rownames(dcanr_input),
  colData = colData
)


dcanr_result <- dcScore(dcanr_input, 
                        colData$Treatment, 
                        dc.method = 'diffcoex'
                        #cor.method = 'spearman'
                        )

gc()
#perform a statistical test: the z-test is selected automatically
raw_p <- dcTest(dcanr_result, dcanr_input, colData$Treatment)
gc()
#adjust p-values (raw p-values from dcTest should NOT be modified)
adj_p <- dcAdjust(raw_p, f = p.adjust, method = 'fdr')


library(igraph)

#get the differential network
dcnet <- dcNetwork(dcanr_result, adj_p)
plot(dcnet, vertex.label = '')

#convert to an adjacency matrix
adjmat <- as_adj(dcnet, sparse = FALSE)
print(adjmat[1:5, 1:5])
#convert to a data.frame
edgedf <- as_data_frame(dcnet, what = 'edges')
print(head(edgedf))


rm(dcanr_result, raw_p, adj_p)
dcMethods()

```




```{r lionessR}
#BiocManager::install("lionessR")

library(lionessR)
library(igraph)
library(reshape2)
library(limma)
library(SummarizedExperiment)

lioness_input <- assay(DeSeq2_return$vsd) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"
      ),
      .before = 1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>%
  column_to_rownames("hgnc_symbol")


colData <-  SummarizedExperiment::colData(DeSeq2_return$vsd) %>%
  as.data.frame() %>% 
  dplyr::select(deseq2_design_condition) %>%
  rownames_to_column("SampleName")

se <- SummarizedExperiment(
  assays = lioness_input %>% as.matrix(),
  rowData = rownames(lioness_input),
  colData = colData
)


nsel=5000
cvar <- apply(assay(se), 1, sd)
dat <- se[tail(order(cvar), nsel), ]


#Next, we will make two condition-specific networks, one for the short-term and one for the long-term metastasis-free survival group. 
netyes <- cor(t(assay(dat)[, dat[[deseq2_design_condition]] == Test]))
netno  <- cor(t(assay(dat)[, dat[[deseq2_design_condition]] == Control]))
netdiff <- netyes-netno



#We use R packages igraph and reshape2 to convert these adjacency matrices to edgelists. 
cormat2 <- rep(1:nsel, each=nsel)
cormat1 <- rep(1:nsel,nsel)
el <- cbind(cormat1, cormat2, c(netdiff))
melted <- melt(upper.tri(netdiff))
melted <- melted[which(melted$value),]
values <- netdiff[which(upper.tri(netdiff))]
melted <- cbind(melted[,1:2], values)
genes <- row.names(netdiff)
melted[,1] <- genes[melted[,1]]
melted[,2] <- genes[melted[,2]]
row.names(melted) <- paste(melted[,1], melted[,2], sep="_")
tosub <- melted
tosel <- row.names(tosub[which(abs(tosub[,3])>0.5),])


# Next, weâ€™ll model the single-sample networks based on co-expression using lionessR. Note that, depending on the size of the dataset, this could take some time to run. 


cormat <- lioness(dat, netFun)
corsub <- assay(cormat[which(row.names(cormat) %in% tosel), ])



#We then run a LIMMA analysis on these edges:

group <- factor(se[[deseq2_design_condition]])
design <- model.matrix(~0+group)
cont.matrix <- makeContrasts(yesvsno = (as.symbol(paste0("group", Test)) - as.symbol(paste0("group", Control))), levels = design)
cont.matrix <- makeContrasts(yesvsno = (groupCarfilzomib - groupWT), levels = design)  
fit <- lmFit(corsub, design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2e <- eBayes(fit2)
toptable <- topTable(fit2e, number=nrow(corsub), adjust="fdr")


#We select the top 50 most differentially co-expressed edges and convert them into an igraph graph.data.frame object for visualization. 

ntop_edges <- 50
toptable_edges <- t(matrix(unlist(c(strsplit(row.names(toptable), "_"))),2))
z <- cbind(toptable_edges[1:ntop_edges,], toptable$logFC[1:ntop_edges])
g <- graph.data.frame(z, directed=FALSE)
E(g)$weight <- as.numeric(z[,3])
E(g)$color[E(g)$weight<0] <- "blue"
E(g)$color[E(g)$weight>0] <- "red"
E(g)$weight <- 1



#Next, we perform a LIMMA analysis on gene expression so that we can also color nodes based on their differential expression:

topgeneslist <- unique(c(toptable_edges[1:ntop_edges,]))
fit <- lmFit(lioness_input, design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2e <- eBayes(fit2)
topDE <- topTable(fit2e, number=nrow(lioness_input), adjust="fdr")
topDE <- topDE[which(row.names(topDE) %in% topgeneslist),]
topgenesDE <- cbind(row.names(topDE), topDE$t)

#We color nodes based on the t-statistic from the LIMMA analysis:

# add t-statistic to network nodes
nodeorder <- cbind(V(g)$name, 1:length(V(g)))
nodes <- merge(nodeorder, topgenesDE, by.x=1, by.y=1)
nodes <- nodes[order(as.numeric(as.character(nodes[,2]))),]
nodes[,3] <- as.numeric(as.character(nodes[,3]))
nodes <- nodes[,-2]
V(g)$weight <- nodes[,2]

# make a color palette
mypalette4 <- colorRampPalette(c("blue","white","white","red"), space="Lab")(256) 
breaks2a <- seq(min(V(g)$weight), 0, length.out=128)
breaks2b <- seq(0.00001, max(V(g)$weight)+0.1,length.out=128)
breaks4 <- c(breaks2a,breaks2b)

# select bins for colors
bincol <- rep(NA, length(V(g)))
for(i in 1:length(V(g))){
    bincol[i] <- min(which(breaks4>V(g)$weight[i]))
}
bincol <- mypalette4[bincol]

# add colors to nodes
V(g)$color <- bincol

#Finally, we visualize these results in a network diagram. In this diagram, edges are colored based on whether they have higher weights in patients with poor (red) or better (blue) MFS. Thicker edges represent higher fold changes. Nodes (genes) are colored based on the t-statistic from a differential expression analysis. Nodes with absolute t-statistic < 1.5 are shown in white, nodes in red/blue have higher expression in patients with poor/better MFS, respectively. 

par(mar=c(0,0,0,0))
plot(g, vertex.label.cex=0.7, vertex.size=10, vertex.label.color = "black", vertex.label.font=3, edge.width=10*(abs(as.numeric(z[,3]))-0.7), vertex.color=V(g)$color)




rm(toptable_edges, toptable)
```




# Using data from STRING 
If you only have RNAseq results, you may want to try if GeneNetworkBuilder could build a network for differential expressed gene. Here is the sample code for that.

```{r STRING}

library(GeneNetworkBuilder)
library(knitr)
library(STRINGdb)
library(org.Hs.eg.db)
library(tidyverse)

GeneNetworkBuilder_input <- paste0(results_dir, "deseq2_results_annotated.tsv") %>% read.delim() %>% dplyr::select(pvalue = padj, logFC = log2FoldChange, gene = hgnc_symbol)



gR <- NULL
try({ ## just in case STRINGdb not work
    library(STRINGdb)
    curr_version_table <- 
      read.table(url("https://string-db.org/api/tsv-no-header/version"),
                 colClasses = "character")$V1[1]
    string_db <- STRINGdb$new( version=curr_version_table, species=9606,
                           score_threshold=400)
    data(diff_exp_example1)
    example1_mapped <- string_db$map(GeneNetworkBuilder_input, "gene", removeUnmappedRows = TRUE )
    i <- string_db$get_interactions(example1_mapped$STRING_id)
    colnames(example1_mapped) <- c("gene", "P.Value", "logFC", "symbols")
    ## get significant up regulated genes.
    genes <- unique(example1_mapped$symbols[example1_mapped$P.Value<0.005 & example1_mapped$logFC>3])
    ### rooted network, guess the root by connections
    x<-networkFromGenes(genes = genes, interactionmap=i, level=3)
    ## filter network
    ## unique expression data by symbols column
    expressionData <- uniqueExprsData(example1_mapped, 
                                       method = 'Max', 
                                       condenseName = "logFC")
    ## merge binding table with expression data by symbols column
    cifNetwork<-filterNetwork(rootgene=x$rootgene, 
                              sifNetwork=x$sifNetwork, 
                              exprsData=expressionData, mergeBy="symbols",
                              miRNAlist=character(0), 
                              tolerance=1, cutoffPVal=0.001, cutoffLFC=1)
    ## convert the id back to symbol
    IDsMap <- expressionData$gene
    names(IDsMap) <- expressionData$symbols
    cifNetwork <- convertID(cifNetwork, IDsMap)
    ## add additional info for searching, any character content columns
    cifNetwork$info1 <- sample(c("groupA", "groupB"),
                               size = nrow(cifNetwork),
                               replace = TRUE)
    cifNetwork$info2 <- sample(c(FALSE, TRUE),
                               size = nrow(cifNetwork),
                               replace = TRUE)
    cifNetwork$info3 <- sample(seq.int(7),
                               size = nrow(cifNetwork),
                               replace = TRUE)
    ## polish network
    gR<-polishNetwork(cifNetwork)
    ## browse network
    browseNetwork(gR)
    
    ## try predefined colors
    cifNetwork$color <- sample(rainbow(7), nrow(cifNetwork), replace = TRUE)
    ## polish network
    gR<-polishNetwork(cifNetwork, nodecolor="color")
    ## browse network
    browseNetwork(gR)
    
    ### unrooted network
    x<-networkFromGenes(genes = genes, interactionmap=i, unrooted=TRUE)
    ## filter network
    ## unique expression data by symbols column
    expressionData <- uniqueExprsData(example1_mapped, 
                                       method = 'Max', 
                                       condenseName = "logFC")
    ## merge binding table with expression data by symbols column
    cifNetwork<-filterNetwork(sifNetwork=x$sifNetwork, 
                              exprsData=expressionData, mergeBy="symbols",
                              miRNAlist=character(0), 
                              tolerance=1, cutoffPVal=0.001,
                              cutoffLFC=1) # set minify=FALSE to retrieve all the interactions
    ## convert the id to symbol
    IDsMap <- expressionData$gene
    names(IDsMap) <- expressionData$symbols
    cifNetwork <- convertID(cifNetwork, IDsMap)
    ## polish network
    gR<-polishNetwork(cifNetwork)
    ## browse network
    browseNetwork(gR)
})
```

### Subset the network by gene list

If you are only interested in a subset of gene list such as genes involved in one gene ontology, you can subset the graph.

```{r, subset}
if(!is.null(gR)){
  library(org.Hs.eg.db)
  goGenes <- mget("GO:0002274", org.Hs.egGO2ALLEGS)[[1]]
  goGenes <- unique(unlist(mget(unique(goGenes), org.Hs.egSYMBOL)))
  gRs <- subsetNetwork(gR, goGenes)
  browseNetwork(gRs)
}
```


```{r david_cola}

#BiocManager::install("cola", update=FALSE)
library(cola)

david_enrichment_data <- 
  results_annotated %>% dplyr::filter(padj < 0.05, log2FoldChange > 1) %>%
  dplyr::filter(!is.na(hgnc_symbol) & !duplicated(hgnc_symbol))%>%
  dplyr::pull(ensembl_gene_id)

david_enrichment <- david_enrichment(
    genes = david_enrichment_data,
    email = "anandagopal.srinivasan@ndorms.ox.ac.uk",
    catalog = c("KEGG_PATHWAY"),
    idtype = "ENSEMBL_GENE_ID",
    species = "Homo sapiens")



```



### Corto
```{r corto, eval=FALSE, include=FALSE}

#install.packages("corto")
invisible(gc())
Corto_input <- 
  #inmat %>%
  assay(DeSeq2_return$vsd) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"
      ),
      .before = 1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>%
  column_to_rownames("hgnc_symbol") %>% 
  as.matrix()

load(system.file("extdata","centroids.rda",package="corto"))

centroids <- mapIds(
        EnsDb.Hsapiens.v86,
        keys = centroids,
        column = "GENEID",
        keytype = "SYMBOL"
      )

TF_list <- read.csv("TFDB.csv") %>%
  dplyr::filter(Is.TF. == "Yes")
TF_list <- TF_list$HGNC.symbol


invisible(gc())
corto_regulon <- corto(Corto_input,
                  centroids=TF_list,
                  nbootstraps=100,
                  p=1e-7,
                  nthreads=2)



corto_regulon[1:10]

```

### Kboost - tbd
```{r kboost, eval=FALSE, include=FALSE}
library("KBoost")

kboost_input <- assay(DeSeq2_return$vsd) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"
      ),
      .before = 1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>%
  column_to_rownames("hgnc_symbol") %>% 
  as.matrix() %>% t()
  
kboost_grn <- KBoost_human_symbol(kboost_input)

```



```{r}
BiocManager::install("AMOUNTAIN", update=FALSE)

library(trena)
library(DIANE)
library(MODA)
library("MODifieR")
```







```{r}

#BioCor
#GNET2
#TRENA
# ComHub python only
#netZooR
#remotes::install_github("OceaneCsn/DIANE", upgrade = "never")

#library(DIANE)
#DIANE::run_app()

hypoxiaGeneSet <- HS_HALLMARK <- msigdbr(species = "Homo sapiens", category = "H")%>% filter(str_detect(gs_name, "HYPOXIA")) %>% dplyr::select(human_gene_symbol, human_ensembl_gene)


hypoxiaGeneSet %>%
  filter(human_gene_symbol %in% DownSigGenes)



```




```{r pathfindR, fig.height=6, fig.width=12}
#install.packages("pak") # if you have not installed "pak"
# need to have java 8 or higher
#https://cimentadaj.github.io/blog/2018-05-25-installing-rjava-on-windows-10/installing-rjava-on-windows-10/index.html
#install.packages("pathfindR")
library(pathfindR)
res <- paste0(results_dir, "deseq2_results_annotated.tsv")
results_annotated <- read.delim(res)
pr_input <- results_annotated %>%
  dplyr::select(Gene.symbol = hgnc_symbol, 
         logFC	= log2FoldChange, 
         adj.P.Val = padj) %>% 
  dplyr::filter(!is.na(adj.P.Val), !is.na(Gene.symbol), !is.na(logFC))


gc()

pathfindR_output_Biogrid_Reactome <- run_pathfindR(
  pr_input,
  gene_sets = "Reactome",
  pin_name_path = "Biogrid",
  min_gset_size = 5,
  max_gset_size = 300,
  output_dir = file.path(results_dir, "pathfindR_results"),
  visualize_enriched_terms = TRUE,
  plot_enrichment_chart = FALSE
)

saveRDS(pathfindR_output_Biogrid_Reactome, file="pathfindR_output_Biogrid_Reactome.RDS")
start <- Sys.time()
pathfindR_clustered <- cluster_enriched_terms(pathfindR_output_Biogrid_Reactome, plot_dend = FALSE, plot_clusters_graph = FALSE, method = "fuzzy")


start <- Sys.time()
pathfindR_clustered_hierarchical <- cluster_enriched_terms(pathfindR_output_Biogrid_Reactome[1:250, ], plot_dend = FALSE, plot_clusters_graph = FALSE)
as.period(Sys.time() - start) %>% round()

pathfindR::term_gene_heatmap(pathfindR_output_Biogrid_Reactome)
enrichment_chart(result_df = pathfindR_output_Biogrid_Reactome, 
                 top_terms = 15)


exp_mat <- assay(DeSeq2_return$dds) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"
      ),
      .before = 1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>%
  column_to_rownames("hgnc_symbol") %>% as.matrix()



enrichment_table <- list(pathfindR_output_Biogrid_Reactome %>% head(30), pathfindR_clustered[pathfindR_clustered$Status == "Representative", ] %>% 
  distinct() %>% group_by(ID)  %>% mutate(count = n()) %>% dplyr::filter(count > 5) %>% dplyr::select(-count, -Cluster) %>% distinct(), 
pathfindR_clustered_hierarchical[pathfindR_clustered_hierarchical$Status == "Representative", ])
score_matrix <- score_terms(enrichment_table = enrichment_table[[3]],
                            exp_mat = exp_mat,
                            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            use_description = TRUE, # default FALSE
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"
```



```{r xgr, fig.height=10, fig.width=10}
library(XGR)
# Specify the location of built-in data
RData.location <- "http://galahad.well.ox.ac.uk/bigdata"
# Load differential expression analysis results

background <- results_annotated$hgnc_symbol
data <- results_annotated %>% 
  filter(log2FoldChange < -1, padj < 0.05) %>%
  pull(hgnc_symbol)

eTerm_Bort_WT <- xEnricherGenes(data=data, background=background, ontology="DO", ontology.algorithm="none", RData.location=RData.location)


xEnrichViewer(eTerm_Bort_WT, 10)

bp_Bort_WT <- xEnrichBarplot(eTerm_Bort_WT, top_num="auto", displayBy="fdr")
bp_Bort_WT

xEnrichDAGplot(eTerm_Bort_WT, top_num="auto", displayBy="fdr", node.info=c("both"), graph.node.attrs=list(fontsize=30), newpage=F)


eTerm_Bort_WT_lea <- xEnricherGenes(data=data, background=background, ontology="DO", ontology.algorithm="lea", RData.location=RData.location)
xEnrichViewer(eTerm_Bort_WT_lea, 10)



list_eTerm <- list(eTerm_Bort_WT, eTerm_Bort_WT_lea)
names(list_eTerm) <- c('DO Tree (-)', 'DO Tree (+)')
bp_Bort_WT_DO <- xEnrichCompare(list_eTerm, displayBy="fc")
bp_Bort_WT_DO + theme(axis.text.y=element_text(size=10))
```


```{r}
TFDB <- read.csv("TFDB.csv") %>% filter(Is.TF. == "Yes")

tfd <- read.delim("C:\\Users\\asrinivasan\\Downloads\\TF_link.tsv.gz")

tfd_dorothea <- dorothea::dorothea_hs
unique_TF_from_TF_link <- tfd %>% dplyr::select(Name.TF) %>% distinct() %>% pull()
unique_TF_from_dorothea <- tfd_dorothea %>% 
  filter(tf %in% TFDB$HGNC.symbol)

tfd %>% filter(!Name.TF %in% TFDB$HGNC.symbol) %>% pull(Name.TF) %>% unique() %>% length()


hTFtarget <- read.delim("C:\\Users\\asrinivasan\\Downloads\\hTFtarget.txt") %>%
  select(-tissue) %>% distinct()

unique_TF_from_hTFtarget <- hTFtarget %>% dplyr::select(TF) %>% distinct() 

unique_TF_from_hTFtarget %>% filter(TF %in% unique_TF_from_TF_link) %>% nrow()

gmt_fname <- system.file("extdata", "pathways.gmt", package = "CEMiTool")
gmt_in <- read_gmt(gmt_fname)

int_df <- read.delim(system.file("extdata", "interactions.tsv", package = "CEMiTool"))

int_df$gene2symbol %>% unique() %>% length()

```


### CEMiTool
```{r CEMiTool, eval=FALSE, include=FALSE}
start <- Sys.time()
#BiocManager::install("CEMiTool")
invisible(gc())
library(CEMiTool)
library(tidyverse)
library(magrittr)
library(SummarizedExperiment)
# Use VST values as input to CEMiTool
cemitool_input <- assay(DeSeq2_return$vsd) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"
      ),
      .before = 1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>%
  column_to_rownames("hgnc_symbol") 


colData <-  SummarizedExperiment::colData(DeSeq2_return$vsd) %>%
  as.data.frame() %>% 
  dplyr::select(deseq2_design_condition) %>%
  rownames_to_column("SampleName")

gmt_fname <- system.file("extdata", "pathways.gmt", package = "CEMiTool")
gmt_in <- read_gmt(gmt_fname)

int_df <- read.delim(system.file("extdata", "interactions.tsv", package = "CEMiTool"))

# Run CEMiTool
cemitool_results <- cemitool(cemitool_input, 
                             colData, 
                             apply_vst = FALSE, 
                             gmt = gmt_in, 
                             interactions = int_df, 
                             verbose = TRUE, 
                             force_beta = TRUE, 
                             class_column = deseq2_design_condition,
                             network_type = "signed")


invisible(gc())
generate_report(cemitool_results, force=TRUE, directory = file.path(results_dir, "CEMItool_report"))


print(Sys.time() - start)

```


```{r regenrich}
#BiocManager::install("RegEnrich")
gc()
library(RegEnrich)
regenrich_input <- assay(DeSeq2_return$dds) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"
      ),
      .before = 1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>%
  column_to_rownames("hgnc_symbol") 

colData <-  SummarizedExperiment::colData(DeSeq2_return$vsd) %>%
  as.data.frame() %>% 
  dplyr::select(deseq2_design_condition) %>%
  rownames_to_column("SampleName")

data(TFs, package = "RegEnrich")
Regenrich_TFs <- TFs
rm(TFs)

Regenrich_object = RegenrichSet(expr = regenrich_input, # expression data (matrix)
                      colData = colData, # sample information (data frame)
                      reg = Regenrich_TFs$TF_name, # regulators
                      method = "Wald_DESeq2", # differentila expression analysis method
                      design = ~ Treatment, # desing model matrix
                      contrast = c(deseq2_design_condition, Test, Control), # contrast
                      networkConstruction = "COEN", # network inference method
                      #proportion = 0.01,
                      enrichTest = "FET") # enrichment analysis method


Regenrich_results = regenrich_diffExpr(Regenrich_object) %>% 
  regenrich_network() %>% 
  regenrich_enrich() %>% 
  regenrich_rankScore() %>%
  results_score()




### not run
library(BiocParallel)
# on non-Windows computers (use 2 workers)
bpparam = register(MulticoreParam(workers = 4, RNGseed = 1234))
# on Windows computers (use 2 workers)
#bpparam = register(SnowParam(workers = 2, RNGseed = 1234))

gc()
object3 = 
  regenrich_diffExpr(Regenrich_object) %>% 
  regenrich_network(networkConstruction = "GRN",
                            BPPARAM = bpparam, minR = 0.3) 

# Obtaining results
#Regenrich_results = results_score(object)
gc()
Regenrich_results_GRN = object3 %>%
  regenrich_enrich() %>% 
  regenrich_rankScore() %>%
  results_score()
```



```{r describe}

all_genes_in_sample <- read.delim(file.path(results_dir, "Car_WT_results_all_genes.tsv")) %>%
  mutate(GENEBIOTYPE = mapIds(#org.Hs.eg.db,
                         EnsDb.Hsapiens.v86,
                         keys=ensembl_gene_id,
                         column="GENEBIOTYPE",
                         keytype="GENEID",
                         multiVals="first"))

all_genes_in_sample %<>%
  filter(padj < 0.05)

all_genes_in_sample %>% filter(ENTREZID == "") %>% nrow()

a <-
results_annotated %>%
  filter(padj < 0.05, abs(log2FoldChange) > 1) %>% 
  mutate(GENEBIOTYPE = mapIds(#org.Hs.eg.db,
                         EnsDb.Hsapiens.v86,
                         keys=ensembl_gene_id,
                         column="GENEBIOTYPE",
                         keytype="GENEID",
                         multiVals="first"))

a %>% filter(GENEBIOTYPE == "protein_coding") %>% 
#  filter(!GENEBIOTYPE == "TEC") %>%
 # filter(!GENEBIOTYPE %>% str_detect("pseudo")) %>%
  View()




##HALLMARKS GENE SETS
  HS_HALLMARK <- msigdbr(species = "Homo sapiens", category = "H")
  HS_HALLMARK <- split(x = HS_HALLMARK$gene_symbol, f = HS_HALLMARK$gs_name)


 
 
Volcano$data %<>% filter(hgnc_symbol %in%  HS_HALLMARK$HALLMARK_INTERFERON_ALPHA_RESPONSE) 


Volcano  
```



```{r yaml, echo=TRUE,warning=FALSE,message=FALSE,error=FALSE, include=FALSE}
library(AnnotationDbi)
library(org.Hs.eg.db)
library(biomaRt)
library("EnsDb.Hsapiens.v86")
library(tidyverse)
library(yaml)
library(tximport)
library(tximportData)

library(ensembldb)
#BiocManager::install("tximportData", update = FALSE)
params <- read_yaml("config.yml")


```


```{r input, echo=TRUE,warning=FALSE,message=FALSE,error=FALSE, include=FALSE}

if(params$kallisto){
  print("kallisto input detected")
  
  #design_files <- list.files(path='designs_kallisto/', pattern = "design_")
  
  # load in the sample metadata file
  sample_meta_data <- list.files(pattern = "*.tsv", path= ".", full.names = TRUE) %>% stringr::str_subset("design") %>% stringr::str_subset("kallisto")
  sample_table <- read.delim(sample_meta_data) %>% dplyr::rename("Sample" = 1)
  
 
  if (dir.exists(params$kallisto_dir)) {
    path <- file.path(params$kallisto_dir, sample_table$Sample, "abundance.h5")
    sample_table <- dplyr::mutate(sample_table, path = path)

  if(params$species == "human"){
    # Create the tx2gene file that will map transcripts to genes
    Tx <- transcripts(EnsDb.Hsapiens.v86, columns=c("tx_id", "gene_id", "symbol"))
    Tx <- as_tibble(Tx)
    Tx <- dplyr::rename(Tx, target_id = tx_id, ens_gene = gene_id, ext_gene = symbol)
    Tx <- dplyr::select(Tx, "target_id", "ens_gene", "ext_gene")
    
  }else if(params$species == "mouse"){
    Tx <- transcripts(EnsDb.Mmusculus.v79, columns=c("tx_id", "gene_id", "symbol"))
    Tx <- as_tibble(Tx)
    Tx <- dplyr::rename(Tx, target_id = tx_id, ens_gene = gene_id, ext_gene = symbol)
    Tx <- dplyr::select(Tx, "target_id", "ens_gene", "ext_gene")  
  }else if(params$species == "macaque"){
    orgSymbols <- keys(org.Mmu.eg.db, keytype="ENSEMBL")
    mart <- useMart(dataset = "mmulatta_gene_ensembl", biomart='ensembl')
    tx2gene <- getBM(attributes = c('ensembl_gene_id', 'ensembl_gene_id_version', 'ensembl_transcript_id', 'ensembl_transcript_id_version','entrezgene_id'),
      filters = 'ensembl_gene_id',
      values = orgSymbols, 
      mart = mart)
    Tx <- tx2gene %>%   dplyr::select(ensembl_transcript_id, ensembl_gene_id)
    colnames(Tx) <- c("TXNAME", "GENEID")

  }else if(params$species == "pig"){
    ensembl = useMart(biomart="ENSEMBL_MART_ENSEMBL",
                  dataset="sscrofa_gene_ensembl", 
                  host="uswest.ensembl.org",
                  ensemblRedirect = FALSE)
    orgSymbols <- unlist(getBM(attributes = 'ensembl_gene_id', mart=ensembl))
    mart <- useMart(dataset = "sscrofa_gene_ensembl", biomart='ensembl', host="uswest.ensembl.org")
    tx2gene <- getBM(attributes = c('ensembl_gene_id', 'ensembl_gene_id_version', 'ensembl_transcript_id', 'ensembl_transcript_id_version','entrezgene_id'),
      filters = 'ensembl_gene_id',
      values = orgSymbols, 
      mart = mart)
    Tx <- tx2gene %>%   dplyr::select(ensembl_transcript_id, ensembl_gene_id)
    colnames(Tx) <- c("TXNAME", "GENEID")
  }else if(params$species == "rabbit"){
    ensembl = useMart(biomart="ENSEMBL_MART_ENSEMBL",
                  dataset="ocuniculus_gene_ensembl", 
                  host="uswest.ensembl.org",
                  ensemblRedirect = FALSE)
    orgSymbols <- unlist(getBM(attributes = 'ensembl_gene_id', mart=ensembl))
    mart <- useMart(dataset = "ocuniculus_gene_ensembl", biomart='ensembl', host="uswest.ensembl.org")
    tx2gene <- getBM(attributes = c('ensembl_gene_id', 'ensembl_gene_id_version', 'ensembl_transcript_id', 'ensembl_transcript_id_version','entrezgene_id'),
      filters = 'ensembl_gene_id',
      values = orgSymbols, 
      mart = mart)
    Tx <- tx2gene %>%   dplyr::select(ensembl_transcript_id, ensembl_gene_id)
    colnames(Tx) <- c("TXNAME", "GENEID")
    }else{ print('please supply a valid species within the config.yml file')}
    
    # import Kallisto transcript counts into R using Tximport
    Txi_gene <-  tximport(path, 
                     type = "kallisto", 
                     tx2gene = Tx, 
                     txOut = FALSE, # TRUE outputs transcripts, FALSE outputs gene-level data
                     countsFromAbundance = "lengthScaledTPM",
                     ignoreTxVersion = TRUE)
    # Write the counts to an object (used for metadata and clustering)
    df_mRNA <- Txi_gene$counts %>% 
            round() %>% 
            data.frame()
    colnames(df_mRNA) <- sample_table$Sample
    
    meta_data <- sample_table
    rownames(meta_data) <- meta_data$Sample
    
  # if (file.exists(paste0('designs_kallisto/',design_files[1]))) {
  #   for (i in design_files){
  #   meta_data <- read.table(paste0('designs_kallisto/',i), sep=",", header = TRUE)
  #   rownames(meta_data) <- meta_data$Sample
  #   df_mRNA_tmp = df_mRNA[,rownames(meta_data)]
  #   all(rownames(meta_data) %in% colnames(df_mRNA_tmp))
  #   assign(paste("meta_data", i, sep = "."), meta_data)}}
  #   else {
  #     print("No design files were detected please add a file called design_<test>_<control>_<test>_<column>.csv. Please refer to documentation on github for more ifnormation")
  #     }
    
  } else {
  print("Please add path to the kallisto dir in the config.yml file as it seems to be missing")
}
} else{


design_files <- list.files(path='designs_featurecounts/', pattern = "design_")
if (file.exists("featurecounts.tsv.gz")) {
  df_mRNA <- read.table(gzfile("featurecounts.tsv.gz"), sep = "\t", header = TRUE, row.names = 1)
  colnames(df_mRNA) <- gsub(".", "-", x = colnames(df_mRNA), fixed = T)
} else {
  print("Please add featurecounts.tsv.gz into the project folder as it seems to be missing")
}


if (file.exists(paste0('designs_featurecounts/',design_files[1]))) {
  for (i in design_files){
  meta_data <- read.table(paste0('designs_featurecounts/',i), sep=",", header = TRUE) 
  rownames(meta_data) <- meta_data$Sample
  df_mRNA_tmp = df_mRNA[,rownames(meta_data)]
  all(rownames(meta_data) %in% colnames(df_mRNA_tmp))
  assign(paste("meta_data", i, sep = "."), meta_data)
  }
} else {
    print("No design files were detected please add a file called design_<test>_<control>_<test>_<column>.csv. Please refer to documentation on github for more ifnormation")
}

}

  write.table(df_mRNA %>% rownames_to_column("id"),
              paste0("df_mRNA.tsv"),
              col.name=TRUE,
              sep="\t",
              na = "NA",
              row.names=FALSE,
              quote=FALSE)
  
  
  
```



```{r}
if (!require("DIANE")) {remotes::install_github("OceaneCsn/DIANE")}

library(DIANE)

DIANE::run_app()
```





















