---
title: "Enrichment Analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = FALSE,
	message = FALSE,
	warning = FALSE
)
```


```{r Source R functions, message=FALSE, warning=FALSE, include=FALSE}

source("R_functions/utils.R", local = knitr::knit_global())
source("R_functions/read_data.R", local = knitr::knit_global())

```


## GSEA

```{r FGSEA table, echo=FALSE, fig.height=10, fig.width=10, message=FALSE, warning=FALSE, paged.print=FALSE}

GSEA <- GSEA_plots(pathways = "kegg", Plot_title, Control, Test, res =  paste0(results_dir, "deseq2_results_annotated.tsv"))

GSEA[[1]]

```

```{r FGSEA plots, echo=FALSE, fig.height=10, fig.width=10, message=FALSE, warning=FALSE, paged.print=FALSE}

GSEA[[2]]


```

# GO Seq

```{r GO seq, fig.height=10, fig.width=8, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}

GoSeq_over <- 
  ORA_GoSeq(title = Plot_title, 
          control =  Control, 
          test =  Test,
          FC = 1,
          #FC = FC_ORA,
          pvalue = pvalue_ORA,
          #res = DeSeq2_results[["res"]],
          #res = "results/deseq2_results_annotated.tsv"
  
  
)

GoSeq_under <- 
  ORA_GoSeq(title = Plot_title, 
          control =  Control, 
          test =  Test,
          FC = -1,
          #FC = FC_ORA,
          pvalue = pvalue_ORA,
          #res = DeSeq2_results[["res"]],
          #res = "results/deseq2_results_annotated.tsv"
  
  
)
```


```{r GoSeq_plot, echo=FALSE, fig.height=10, fig.width=8, message=FALSE, warning=FALSE, paged.print=FALSE}
GoSeq_over
GoSeq_under
```


```{r clusterProfiler, fig.height=15, fig.width=8, message=FALSE, warning=FALSE, include=FALSE, paged.print=TRUE}
invisible(gc())
clusterProfiler_Plots <- clusterProfiler_Plots(
  res = paste0(results_dir, "deseq2_results_annotated.tsv"),
  title = paste(Plot_title, Control, "vs", Test))

```

```{r clusterProfiler_Plots, fig.height=7, fig.width=7}
clusterProfiler_Plots
```


```{r enrichR, include=FALSE}
DeSeq2_return$res %>% nrow()
#invisible(gc())
setEnrichrSite("Enrichr") # Human genes
dbs <- listEnrichrDbs()
dbs <- c("GO_Biological_Process_2021",
         "GO_Cellular_Component_2021",
         "GO_Molecular_Function_2021",
         "Reactome_2022", 
         "ChEA_2022")

res <- paste0(results_dir, "deseq2_results_annotated.tsv")
results_annotated <- read.delim(res)

universe <- results_annotated %>% pull(hgnc_symbol)
UpSigGenes <- results_annotated %>%
    dplyr::filter(padj < 0.05, log2FoldChange > 1
                  #!is.na(ENSEMBL)
    ) %>% pull(hgnc_symbol)
DownSigGenes <- results_annotated %>%
    dplyr::filter(padj < 0.05, log2FoldChange < -1
                  #!is.na(ENSEMBL)
    ) %>% pull(hgnc_symbol)


UPenriched <- enrichr(UpSigGenes, dbs)
DOWNenriched <- enrichr(DownSigGenes, dbs) 

enrichR_plot <- list()
for (i in 1:length(dbs)) {
  enrichR_plot[[i]] <- plotEnrich(UPenriched[[i]], showTerms = 5, numChar = 40, y = "Count", orderBy = "Combined.Score")
  enrichR_plot[[i+length(dbs)]] <- plotEnrich(DOWNenriched[[i]], showTerms = 5, numChar = 100, y = "Count", orderBy = "Combined.Score")
}




```

##Upregulated Pathways
```{r enricher plots, fig.height=5, fig.width=7}


for (i in 1:length(dbs)) {
  enrichR_plot[[i]] <- enrichR_plot[[i]] + scale_x_discrete(label = function(x) str_wrap(x, width = 10))
  print(enrichR_plot[[i]])
  ggsave(filename =  paste0(plots_dir, Plot_title, "_overrepresented_", dbs[i], "_Enricher", "_plot.png"), width = 7, height = 7, units = "in")

}

```


##Downregulated Pathway


```{r enricher plots2, fig.height=5, fig.width=7, message=FALSE, warning=FALSE}

for (i in 1:length(dbs)) {
  enrichR_plot[[i+length(dbs)]] <- enrichR_plot[[i+length(dbs)]] + scale_x_discrete(label = function(x) str_wrap(x, width = 10))
  print(enrichR_plot[[i+length(dbs)]])
  ggsave(filename =  paste0(plots_dir, Plot_title, "_underrepresented_", dbs[i+length(dbs)], "_Enricher", "_plot.png"), width = 7, height = 7, units = "in")
}

```



```{r DAVID, eval=FALSE, include=FALSE}
#BiocManager::install("RDAVIDWebService")

# ## try http:// if https:// URLs are not supported
# source("https://bioconductor.org/biocLite.R")
# biocLite("RDAVIDWebService")
# #install.packages("RDAVIDWebService")
# library(clusterProfiler)
# # devtools::install_github("XiaoFei-Zhao/RDAVIDWebService")
# library(RDAVIDWebService)
# res <- paste0(results_dir, "deseq2_results_annotated.tsv")
# results_annotated <- read.delim(res)
# universe <- results_annotated %>% pull(ENTREZID)
# sigGenes <- results_annotated %>%
#   dplyr::filter(padj < 0.05,
#                 abs(log2FoldChange) > 1
#                   #!is.na(ENSEMBL)
#     ) %>% pull(ENTREZID)
# 
# david = enrichDAVID(gene = sigGenes, idType="ENTREZ_GENE_ID",
#                     universe = universe, annotation="KEGG_PATHWAY")

```


## Aleternative GSE and ORA



```{r ananlysis_Pi, fig.height=7, fig.width=7, message=FALSE, warning=FALSE}

res <- paste0(results_dir, "deseq2_results_annotated.tsv")
results_annotated <- read.delim(res)

print("Upregulated genes")

xgr_ora_plots <- function(res = paste0(results_dir, "deseq2_results_annotated.tsv"), pathways = "DO", FC = 1, pvalue=pvalue_ORA, structured=TRUE) {
  
  results_annotated <- read.delim(res)
  if (FC < 0) {xgr_data <- results_annotated %>%
    dplyr::filter(padj < pvalue_ORA, log2FoldChange < FC, baseMean >25
                  #!is.na(ENSEMBL)
    ) 
  }

  if (FC > 0) {xgr_data <- results_annotated %>%
    dplyr::filter(padj < pvalue_ORA, log2FoldChange > FC, baseMean >25
                  #!is.na(ENSEMBL)
    ) 
  }
  ontology.algorithm <- ifelse(structured, "lea", "none")
  data <- xgr_data %>% 
  dplyr::filter(!is.na(hgnc_symbol) & !duplicated(hgnc_symbol))%>%
  dplyr::pull(hgnc_symbol)
background <- results_annotated %>% 
  dplyr::filter(!is.na(hgnc_symbol) & !duplicated(hgnc_symbol)) %>%
  dplyr::pull(hgnc_symbol)
ontology <- list("KEGG" = "MsigdbC2KEGG", "Reactome" = "MsigdbC2REACTOME", "DO" = "DO", "GOMF" = "GOMF", "GOBP" = "GOBP")
eTerm <- xEnricherGenes(data=data, background=background, ontology= ontology[[pathways]], ontology.algorithm = ontology.algorithm)

eTerm_concise <- xEnrichConciser(eTerm)


bp <- xEnrichBarplot(eTerm, top_num=10, displayBy="adjp")
bp <- bp + scale_x_discrete(labels = function(x) str_wrap(x, width = 40)) 
print(bp)

over_or_under <- ifelse(FC > 0, "over", "under")

ggsave(filename =  paste0(plots_dir, Plot_title, "_", over_or_under, "_represented_", pathways, "_XGR_plot.png"), width = 10, height = 14, units = "in")

return(bp)  
}
#a <- xRDataLoader(RData.customised='JKscience_TS1A', RData.location=RData.location)
xgr_reactome_up <- xgr_ora_plots(res, "Reactome", 1, pvalue_ORA, FALSE)
xgr_do_up <- xgr_ora_plots(res, "DO", 1, pvalue_ORA, TRUE)
xgr_kegg_up <- xgr_ora_plots(res, "KEGG", 1, pvalue_ORA, FALSE)
xgr_gobp_up <- xgr_ora_plots(res, "GOBP", 1, pvalue_ORA, TRUE)
xgr_gomf_up <- xgr_ora_plots(res, "GOMF", 1, pvalue_ORA, TRUE)

print("DOWN regulated genes")

xgr_reactome_down <- xgr_ora_plots(res, "Reactome", -1, pvalue_ORA, FALSE)
xgr_do_down <- xgr_ora_plots(res, "DO", -1, pvalue_ORA, TRUE)
xgr_kegg_down <- xgr_ora_plots(res, "KEGG", -1, pvalue_ORA, FALSE)
xgr_gomf_down <- xgr_ora_plots(res, "GOMF", -1, pvalue_ORA, TRUE)
xgr_gobp_down <- xgr_ora_plots(res, "GOBP", -1, pvalue_ORA, TRUE)

```



```{r Pi, fig.height=7, fig.width=7, eval=FALSE}
#xDefineOntology

## below works tuned off for now. Review later.
# xEnrichDAGplot(eTerm, top_num="auto", displayBy="fdr", node.info=c("both"), graph.node.attrs=list(fontsize=30), newpage=F)
# 
# RData.location <- "http://galahad.well.ox.ac.uk/bigdata"
# network_input <- results_annotated %>%
#   dplyr::filter(!is.na(hgnc_symbol),
#                 !duplicated(hgnc_symbol),
#                 log2FoldChange < -1) %>%
#   dplyr::select(hgnc_symbol, padj)
# 
# subg_func <- xSubneterGenes(data=network_input, network="STRING_low", subnet.size=75, RData.location=RData.location)
# subg <- subg_func
# pattern <- -log10(as.numeric(V(subg)$significance))
# xVisNet(g=subg, pattern=pattern, vertex.shape="sphere", vertex.label.font=2, newpage=F)


# RData.location <- "http://galahad.well.ox.ac.uk/bigdata"
# 
# #### xPier test ####
# 
# xgr_data_up <- results_annotated %>% 
#   dplyr::filter(padj < 0.05, log2FoldChange > 1)
# xgr_data_down <- results_annotated %>% 
#   dplyr::filter(padj < 0.05, log2FoldChange < -1)
# 
# 
# 
# xgr_ranks <- xgr_data_up %>% 
#   dplyr::filter(!is.na(hgnc_symbol), !duplicated(hgnc_symbol)) %>%
#   dplyr::select(hgnc_symbol, stat) %>% mutate(rank = dense_rank(-stat)) %>% dplyr::select(-stat) %>% mutate(rank = rank/nrow(.))
# 
# xgr_ranks <- xgr_data_down %>% 
#   dplyr::filter(!is.na(hgnc_symbol), !duplicated(hgnc_symbol)) %>%
#   dplyr::select(hgnc_symbol, stat) %>% mutate(rank = dense_rank(-stat)) %>% dplyr::select(-stat) %>% mutate(rank = rank/nrow(.))  
#   
#   
# pNode <- xPierGenes(xgr_ranks, network = "STRING_high", restart=0.7)
# 
# 
# ##How does this work???
# xPGSEA <- xPierGSEA(pNode, ontology = "MsigdbC2REACTOME")
# 
# eTerm <- xPierPathways(pNode, priority.top=100, ontology="MsigdbC2REACTOME", RData.location=RData.location)
# eTerm_nonred <- xEnrichConciser(eTerm)
# 
# # view the top pathways/terms
# xEnrichViewer(eTerm_nonred) %>% kableExtra::kable()
# 
# 
# bp <- xEnrichBarplot(eTerm_nonred, top_num="auto", displayBy="fdr", FDR.cutoff=1e-3, wrap.width=50, signature=FALSE)
# bp
# 
# ggsave(paste0(plots_dir, Plot_title,"_XGR_Pi.png"))
```


## TF ANALYSIS

```{r chea3}
# library(httr)
# library(jsonlite)

UpSigGenes <- results_annotated %>%
  drop_na() %>%
  dplyr::filter(padj < 0.05, log2FoldChange > 1
                  #!is.na(ENSEMBL)
    ) %>% pull(hgnc_symbol)


DownSigGenes <- results_annotated %>%
  drop_na() %>%
  dplyr::filter(padj < 0.05, log2FoldChange < -1
                  #!is.na(ENSEMBL)
    ) %>% pull(hgnc_symbol)

url = "https://maayanlab.cloud/chea3/api/enrich/"
encode = "json"
query_name <- paste(Plot_title, Control, "vs", Test, "Upregulated", sep = "_")
payload = list(query_name = query_name, gene_set = UpSigGenes)

#POST to ChEA3 server
response = POST(url = url, body = payload, encode = encode)
json = content(response, "text")

#write.table(sigGenes, "C:\\Users\\asrinivasan\\Desktop\\sigGenes.txt", row.names = FALSE, col.names = FALSE, sep = ",", quote = FALSE)

#results as list of R dataframes
chea3_results_upregulated = fromJSON(json)


#####
query_name <- paste(Plot_title, Control, "vs", Test, "Downregulated", sep = "_")
payload = list(query_name = query_name, gene_set = DownSigGenes)

#POST to ChEA3 server
response = POST(url = url, body = payload, encode = encode)
json = content(response, "text")

#write.table(sigGenes, "C:\\Users\\asrinivasan\\Desktop\\sigGenes.txt", row.names = FALSE, col.names = FALSE, sep = ",", quote = FALSE)

#results as list of R dataframes
chea3_results_downregulated = fromJSON(json)


chea3_top_hits <- bind_rows(chea3_results_upregulated$`Integrated--meanRank` %>% head(50), 
chea3_results_downregulated$`Integrated--meanRank` %>% head(50))

```



```{r decoupleR_TF}
#remotes::install_github('saezlab/OmnipathR')
#BiocManager::install("saezlab/decoupleR")
#BiocManager::install("dorothea")
#BiocManager::install("progeny")
invisible(gc())

# library("dorothea")
# library("decoupleR")

dorothea_df <- dorothea_hs %>%
  dplyr::filter(confidence %in% c("A", "B", "C")) %>%
  dplyr::select(source = tf, confidence, target, mor) %>%
  as.data.frame() %>%
  mutate(weight = mor/match(confidence, LETTERS))

collectri_net <- get_collectri(organism='human', split_complexes=FALSE) %>%
  mutate("weight" = mor)

decoupleR_mat <- 
  DeSeq2_return$vsd %>%
  assay() %>% as.data.frame() %>% 
  rownames_to_column("GeneID") %>%
  mutate(
    hgnc_symbol = mapIds(EnsDb.Hsapiens.v86, keys = GeneID, keytype = "GENEID", column = "SYMBOL"), .before = 1) %>%
   drop_na() %>%
  select(-2) %>%
  column_to_rownames("hgnc_symbol") %>% as.matrix()

## if annotating is done after results from deseq
# decoupleR_mat <-
#   DeSeq2_return$vsd %>% 
#   assay() %>% 
#   as.data.frame() %>% 
#   rownames_to_column("ensembl_gene_id") %>% 
#   left_join(results_annotated %>% dplyr::select(ensembl_gene_id, hgnc_symbol) %>% unique()) %>% dplyr::select(-ensembl_gene_id) %>% distinct(hgnc_symbol, .keep_all = TRUE) %>% drop_na() %>% column_to_rownames("hgnc_symbol") %>% as.matrix()

#decoupleR_net <- get_progeny(organism = 'human', top = 300)
#decoupleR_net <- get_dorothea(organism='human', levels=c('A', 'B', 'C'))

sample_acts <- run_wmean(mat=decoupleR_mat, 
                         net=collectri_net, 
                         .source='source', 
                         .target='target',
                  .mor='weight', 
                  times = 100, 
                  minsize = 5)

### number of tfs for heatmap
n_tfs <- 50

# Transform to wide matrix
sample_acts_mat <- sample_acts %>%
  filter(statistic == 'norm_wmean') %>%
  pivot_wider(id_cols = 'condition', names_from = 'source',
              values_from = 'score') %>%
  column_to_rownames('condition') %>%
  as.matrix()

# Get top tfs with more variable means across clusters
tfs <- sample_acts %>%
  group_by(source) %>%
  summarise(std = sd(score)) %>%
  arrange(-abs(std)) %>%
  head(n_tfs) %>%
  pull(source)

decoupleR_mat_top50_variable_TFs <- tfs
sample_acts_mat <- sample_acts_mat[,tfs]

# Scale per sample
sample_acts_mat <- scale(sample_acts_mat)

# Choose color palette
palette_length = 100
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)

my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 3, length.out=floor(palette_length/2)))


# Plot
pheatmap(sample_acts_mat, border_color = NA, color=my_color, breaks = my_breaks) %>% ggplotify::as.ggplot()
ggsave(paste0(plots_dir, "top50_mat_TFs_decoupleR.png"), width = 10, height = 4, units = "in")

```



```{r decoupleR_TF_Degs}
# Extract t-values per gene

decoupleR_deg <- results_annotated %>% 
  dplyr::select(hgnc_symbol, stat, log2FoldChange, padj) %>% 
  distinct(hgnc_symbol, .keep_all = TRUE) %>% 
  drop_na() %>% 
  column_to_rownames("hgnc_symbol") %>% 
  as.matrix()

contrast_acts <- run_wmean(
  mat = decoupleR_deg[, "stat"],
  net = collectri_net,
  .source = 'source',
  .target = 'target',
  .mor = 'weight',
  times = 100,
  minsize = 5
)


# Filter norm_wmean
f_contrast_acts <- contrast_acts %>%
  filter(statistic == 'norm_wmean') %>%
  mutate(rnk = NA)

# Filter top TFs in both signs
msk <- f_contrast_acts$score > 0
f_contrast_acts[msk, 'rnk'] <- rank(-f_contrast_acts[msk, 'score'])
f_contrast_acts[!msk, 'rnk'] <- rank(-abs(f_contrast_acts[!msk, 'score']))
tfs <- f_contrast_acts %>%
  arrange(rnk) %>%
  head(n_tfs) %>%
  pull(source)
f_contrast_acts <- f_contrast_acts %>%
  filter(source %in% tfs)
decoupleR_deg_top50_variable_TFs <- tfs

# Plot
ggplot(f_contrast_acts, aes(x = reorder(source, score), y = score)) + 
    geom_bar(aes(fill = score), stat = "identity") +
    scale_fill_gradient2(low = "darkblue", high = "indianred", 
        mid = "whitesmoke", midpoint = 0) + 
    theme_pubclean() +
    theme(axis.title = element_text(face = "bold", size = 12),
        axis.text.x = 
            element_text(angle = 45, hjust = 1, size =10, face= "bold"),
        axis.text.y = element_text(size =10, face= "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
    xlab("Pathways")
ggsave(paste0(plots_dir, "top50_deg_TFs_decoupleR.png"), width = 12, units = "in")
#######

decoupleR_vis_TF <- function(tf = NULL,
                             
                             net = collectri_net, 
                             deg = decoupleR_deg) {
  tf <- f_contrast_acts %>% 
  filter(rnk == 1, 
         score == max(score)) %>% 
  pull(source)
  
  df <- net %>%
    filter(source == tf) %>%
    arrange(target) %>%
    mutate(ID = target, color = "3") %>%
    column_to_rownames('target')
  
  inter <- sort(intersect(rownames(deg), rownames(df)))
  df <- df[inter,]
  df[, c('stat', 'log2FoldChange', 'padj')] <- deg[inter,]
  df <- df %>%
    mutate(color = if_else(mor > 0 & stat > 0, '1', color)) %>%
    mutate(color = if_else(mor > 0 & stat < 0, '2', color)) %>%
    mutate(color = if_else(mor < 0 & stat > 0, '2', color)) %>%
    mutate(color = if_else(mor < 0 & stat < 0, '1', color))
  
  plot <-
    ggplot(df, aes(
    x = log2FoldChange,
    y = -log10(padj),
    color = color,
    size = abs(weight)
  )) +
    geom_point() +
    scale_colour_manual(values = c("red", "royalblue3", "grey")) +
    geom_label_repel(aes(label = ID, size = 1)) +
    theme_minimal() +
    theme(legend.position = "none") +
    geom_vline(xintercept = 0, linetype = 'dotted') +
    geom_hline(yintercept = 0, linetype = 'dotted') +
    ggtitle(tf)
  
  return(plot)
  
}


decoupleR_vis_TF()
intersect(decoupleR_deg_top50_variable_TFs, decoupleR_mat_top50_variable_TFs) %>% unique() 

chea3_top_hits %>% filter(TF %in% intersect(decoupleR_deg_top50_variable_TFs, decoupleR_mat_top50_variable_TFs)) %>% pull(TF)
```




```{r decoupleR_PROGENy, echo=FALSE, message=FALSE, warning=FALSE}
#BiocManager::install("OmnipathR")
#BiocManager::install("remotes")
#BiocManager::install("dorothea")
#devtools::install_github("saezlab/progeny")
invisible(gc())


PROGENy_df <-get_progeny(organism = 'human', top = 500)

decoupleR_mat <- 
  DeSeq2_return$vsd %>%
  assay() %>% as.data.frame() %>% 
  rownames_to_column("GeneID") %>%
  mutate(
    hgnc_symbol = mapIds(EnsDb.Hsapiens.v86, keys = GeneID, keytype = "GENEID", column = "SYMBOL"), .before = 1) %>%
   drop_na() %>%
  select(-2) %>%
  column_to_rownames("hgnc_symbol") %>% as.matrix()

## if annotating is done after results from deseq
# decoupleR_mat <-
#   DeSeq2_return$vsd %>% 
#   assay() %>% 
#   as.data.frame() %>% 
#   rownames_to_column("ensembl_gene_id") %>% 
#   left_join(results_annotated %>% dplyr::select(ensembl_gene_id, hgnc_symbol) %>% unique()) %>% dplyr::select(-ensembl_gene_id) %>% distinct(hgnc_symbol, .keep_all = TRUE) %>% drop_na() %>% column_to_rownames("hgnc_symbol") %>% as.matrix()

decoupleR_net <- get_progeny(organism = 'human', top = 500)
#decoupleR_net <- get_dorothea(organism='human', levels=c('A', 'B', 'C'))

sample_acts <- run_wmean(mat=decoupleR_mat, 
                         net=decoupleR_net, 
                         .source='source', 
                         .target='target',
                  .mor='weight', 
                  times = 100, 
                  minsize = 5)

### number of tfs for heatmap
n_tfs <- 25

# Transform to wide matrix
sample_acts_mat <- sample_acts %>%
  filter(statistic == 'norm_wmean') %>%
  pivot_wider(id_cols = 'condition', names_from = 'source',
              values_from = 'score') %>%
  column_to_rownames('condition') %>%
  as.matrix()

# Get top tfs with more variable means across clusters
tfs <- sample_acts %>%
  group_by(source) %>%
  summarise(std = sd(score)) %>%
  arrange(-abs(std)) %>%
  head(n_tfs) %>%
  pull(source)


sample_acts_mat <- sample_acts_mat[,tfs]

# Scale per sample
sample_acts_mat <- scale(sample_acts_mat)

# Choose color palette
palette_length = 100
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)

my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 3, length.out=floor(palette_length/2)))


# Plot
pheatmap(sample_acts_mat, border_color = NA, color=my_color, breaks = my_breaks) 

```



```{r decoupleR_PROGENy_Degs, echo=FALSE, message=FALSE, warning=FALSE}
# Extract t-values per gene

decoupleR_deg <- results_annotated %>% 
  dplyr::select(hgnc_symbol, stat, log2FoldChange, padj) %>% 
  distinct(hgnc_symbol, .keep_all = TRUE) %>% 
  drop_na() %>% 
  column_to_rownames("hgnc_symbol") %>% 
  as.matrix()

contrast_acts <- run_wmean(
  mat = decoupleR_deg[, "stat"],
  net = decoupleR_net,
  .source = 'source',
  .target = 'target',
  .mor = 'weight',
  times = 100,
  minsize = 5
)
# Filter norm_wmean
f_contrast_acts <- contrast_acts %>%
  filter(statistic == 'norm_wmean')


# Filter norm_wmean
f_contrast_acts <- contrast_acts %>%
  filter(statistic == 'norm_wmean') %>%
  mutate(rnk = NA)

# Filter top TFs in both signs
msk <- f_contrast_acts$score > 0
f_contrast_acts[msk, 'rnk'] <- rank(-f_contrast_acts[msk, 'score'])
f_contrast_acts[!msk, 'rnk'] <- rank(-abs(f_contrast_acts[!msk, 'score']))
tfs <- f_contrast_acts %>%
  arrange(rnk) %>%
  head(n_tfs) %>%
  pull(source)
f_contrast_acts <- f_contrast_acts %>%
  filter(source %in% tfs)

# Plot
ggplot(f_contrast_acts, aes(x = reorder(source, score), y = score)) + 
    geom_bar(aes(fill = score), stat = "identity") +
    scale_fill_gradient2(low = "darkblue", high = "indianred", 
        mid = "whitesmoke", midpoint = 0) + 
    theme_minimal() +
    theme(axis.title = element_text(face = "bold", size = 12),
        axis.text.x = 
            element_text(angle = 45, hjust = 1, size =10, face= "bold"),
        axis.text.y = element_text(size =10, face= "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
    xlab("Pathways")


#######

decoupleR_vis_TF <- function(tf = NULL,
                             
                             net = dorothea_df, 
                             deg = decoupleR_deg) {
  tf <- f_contrast_acts %>% 
  filter(rnk == 1, 
         score == min(score)) %>% 
  pull(source)
  
  df <- net %>%
    filter(source == tf) %>%
    arrange(target) %>%
    mutate(ID = target, color = "3") %>%
    column_to_rownames('target')
  
  inter <- sort(intersect(rownames(deg), rownames(df)))
  df <- df[inter,]
  df[, c('stat', 'log2FoldChange', 'padj')] <- deg[inter,]
  df <- df %>%
    mutate(color = if_else(weight > 0 & stat > 0, '1', color)) %>%
    mutate(color = if_else(weight > 0 & stat < 0, '2', color)) %>%
    mutate(color = if_else(weight < 0 & stat > 0, '2', color)) %>%
    mutate(color = if_else(weight < 0 & stat < 0, '1', color))
  
  plot <-
    ggplot(df, aes(
    x = log2FoldChange,
    y = -log10(padj),
    color = color,
    size = abs(weight)
  )) +
    geom_point() +
    scale_colour_manual(values = c("red", "royalblue3", "grey")) +
    geom_label_repel(aes(label = ID, size = 1)) +
    theme_minimal() +
    theme(legend.position = "none") +
    geom_vline(xintercept = 0, linetype = 'dotted') +
    geom_hline(yintercept = 0, linetype = 'dotted') +
    ggtitle(tf)
  
  return(plot)
  
}


decoupleR_vis_TF(net = decoupleR_net)
```



```{r TFEA.ChIP}
#BiocManager::install("TFEA.ChIP")
invisible(gc())
#library(TFEA.ChIP)
TFEA_ChIP_input_raw <- preprocessInputData( DeSeq2_return$res )

#extract vector with names of upregulated genes
Genes.Upreg <- Select_genes( TFEA_ChIP_input_raw, min_LFC = 1 )
#extract vector with names of downregulated genes
Genes.Downreg <- Select_genes( TFEA_ChIP_input_raw, max_LFC = -1 )
#extract vector with names of non-responsive genes
Genes.Control <- Select_genes( TFEA_ChIP_input_raw,
    min_pval = 0.05, max_pval = 1,
    min_LFC = -0.25, max_LFC = 0.25 )

CM_list_UP <- contingency_matrix( Genes.Upreg, Genes.Control ) #generates list of contingency tables, one per dataset
pval_mat_UP <- getCMstats( CM_list_UP ) #generates list of p-values and OR from association test


CM_list_DOWN <- contingency_matrix( Genes.Downreg, Genes.Control ) #generates list of contingency tables, one per dataset
pval_mat_DOWN <- getCMstats( CM_list_DOWN ) #generates list of p-values and OR from association test

print("Upregulated TFs")
TF_ranking_up_gsea <- rankTFs( pval_mat_UP, rankMethod = "gsea", makePlot = TRUE )

TF_ranking_up_wilcoxon <- rankTFs( pval_mat_UP, rankMethod = "wilcoxon", makePlot = TRUE )
TF_ranking_up_wilcoxon[[ "TFranking_plot" ]]


TF_ranking_down_gsea <- rankTFs( pval_mat_DOWN, rankMethod = "gsea", makePlot = TRUE )

print("Downregulated TFs")
TF_ranking_down_wilcoxon <- rankTFs( pval_mat_DOWN, rankMethod = "wilcoxon", makePlot = TRUE )
TF_ranking_down_wilcoxon[[ "TFranking_plot" ]]


TF_ranking_down_wilcoxon %>% head(50) %>% bind_rows(TF_ranking_up_wilcoxon %>% head(50)) %>% filter(TF %in% decoupleR_deg_top50_variable_TFs)


TFEA_Chip_ranked_TFs <- c(bind_rows(TF_ranking_down_gsea$TF_ranking, TF_ranking_up_gsea$TF_ranking) %>% filter(pVal < 0.05) %>% pull(TF) %>% unique(), bind_rows(TF_ranking_down_wilcoxon, TF_ranking_up_wilcoxon) %>% filter(wilc_pval < 0.05)  %>% pull(TF) %>% unique()) %>% unique()

TFEA_chip_GSEA_result <- GSEA_run( TFEA_ChIP_input_raw$Genes, TFEA_ChIP_input_raw$log2FoldChange, get.RES = TRUE) #run GSEA analysis


TFEA_Chip_GSEA_enriched_TFs <- TFEA_chip_GSEA_result$Enrichment.table %>% filter(pval.adj < 0.05) %>% pull(TF) %>% unique()


print("Sigificant enriched TFs in TFEA.Chip method are:")
unique(c(TFEA_Chip_GSEA_enriched_TFs, TFEA_Chip_ranked_TFs))
```


```{r cetf}
library(CeTF)
# Creating a variable with annotation data
CeTF_anno <- as.data.frame(colData(DeSeq2_return$dds)) 



# Creating a variable with count data
CeTF_counts <- assay(DeSeq2_return$dds) %>% 
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"
      ),
      .before = 1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>%
  column_to_rownames("hgnc_symbol") %>% 
  as.matrix()

# Sorting count data samples by conditions (untrt and trt) - this is not needed as deseq2 has already arranged Treatment column of colData as factor with two levels, Control/WT is level 1. 
CeTF_input <- CeTF_counts[, order(CeTF_anno[[deseq2_design_condition]], decreasing = TRUE)]

# Loading the Transcript Factors (TFs) character
data("TFs")
CeTF_TFs <- TFs %>%
  mapIds(
        EnsDb.Hsapiens.v86,
        keys = .,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"
      ) %>% unique()

TFs <- read.csv("TFDB.csv") %>%
  dplyr::filter(Is.TF. == "Yes") %>%
  filter(HGNC.symbol %in% CeTF_TFs) %>% 
  pull(HGNC.symbol)

# Performing the complete analysis
CeTF_out <- runAnalysis(mat = CeTF_input, 
                   conditions=c(Test, Control),
                   lfc = 1,
                   padj = 0.05,
                   TFs = TFs,
                   nSamples1 = CeTF_anno[[deseq2_design_condition]] %>% str_count(Test) %>% sum(),
                   nSamples2= CeTF_anno[[deseq2_design_condition]] %>% str_count(Control) %>% sum(),
                   tolType = "mean",
                   #diffMethod = "DESeq2", #"Reverter", 
                   diffMethod = "Reverter", 
                   data.type = "counts")

RIF_table <- OutputData(CeTF_out) %>%
  filter(abs(RIF1) > 1.96 | abs(RIF2) > 1.96)


# Using the runAnalysis output (CeTF class object)
SmearPlot(object = CeTF_out, 
          #diffMethod = 'DESeq2', #"Reverter", 
          diffMethod = "Reverter", 
          lfc = 1, 
          conditions = c(Test, Control), 
          type = "DE")


netConditionsPlot(CeTF_out)

CeTF_keytfs <- CeTF::NetworkData(CeTF_out, "keytfs")

CeTF_networks <- list()
#names(CeTF_networks) <- c(Test, Control)
CeTF_networks[[Test]] <- unique(c(as.character(NetworkData(CeTF_out, "network1")[, "gene1"]), 
                  as.character(NetworkData(CeTF_out, "network1")[, "gene2"])))

CeTF_networks[[Control]] <- unique(c(as.character(NetworkData(CeTF_out, "network2")[, "gene1"]), 
                  as.character(NetworkData(CeTF_out, "network2")[, "gene2"])))


Volcano_Plots(de_seq = DeSeq2_return$de_seq, 
              test = Test, 
              res = DeSeq2_return$res, 
              labels = TRUE,
              title = "PCIT Network Genes",
              pval = pvalue_VP, 
              label_list = RIF_table$TF,
              filter_specific = c(CeTF_networks[[Test]], CeTF_networks[[Control]]),
              FC = 1)



```

```{r RcisTarget}
## Generates set of motifs based on GeneSets - can use either modules from WGCNA or DEGs

invisible(gc())
# library(RcisTarget)
# library(RcisTarget.hg19.motifDBs.cisbpOnly.500bp)

# Rankings
data(hg19_500bpUpstream_motifRanking_cispbOnly)


# Load gene sets to analyze. e.g.:
hypoxiaGeneSet <- read.table(file.path(system.file('examples', package='RcisTarget'), "hypoxiaGeneSet.txt"), stringsAsFactors=FALSE)[,1]


RcisT_geneLists <- list(CarUp = UpSigGenes, CarDown = DownSigGenes)

# Select motif database to use (i.e. organism and distance around TSS)
data(motifAnnotations_hgnc)


motifRankings <- importRankings("../hg38_10kbp_up_10kbp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather")

# Motif enrichment analysis:
#motifEnrichmentTable_wGenes <- cisTarget(RcisT_geneLists, hg19_500bpUpstream_motifRanking_cispbOnly,
#                               motifAnnot=motifAnnotations)

motifEnrichmentTable_wGenes <- cisTarget(RcisT_geneLists, motifRankings,
                               motifAnnot=motifAnnotations)


RcisTarget_TFs <- motifEnrichmentTable_wGenes$TF_highConf %>% str_replace_all(" \\s*\\([^\\)]+\\). ", "") %>%
  str_split("; ") %>% unlist() %>% unique()



invisible(gc())
rm(hg19_500bpUpstream_motifRanking_cispbOnly)

```




```{r pathfindR_reactome, fig.height=6, fig.width=12}
#install.packages("pak") # if you have not installed "pak"
# need to have java 8 or higher
#https://cimentadaj.github.io/blog/2018-05-25-installing-rjava-on-windows-10/installing-rjava-on-windows-10/index.html
#install.packages("pathfindR")
library(pathfindR)
res <- paste0(results_dir, "deseq2_results_annotated.tsv")
results_annotated <- read.delim(res)
pr_input <- results_annotated %>%
  dplyr::select(Gene.symbol = hgnc_symbol, 
         logFC	= log2FoldChange, 
         adj.P.Val = padj) %>% 
  dplyr::filter(!is.na(adj.P.Val), !is.na(Gene.symbol), !is.na(logFC))


gc()

pathfindR_output_Biogrid_Reactome <- run_pathfindR(
  pr_input,
  gene_sets = "Reactome",
  pin_name_path = "Biogrid",
  min_gset_size = 5,
  max_gset_size = 300,
  output_dir = file.path(results_dir, "pathfindR_results_Reactome"),
  #visualize_enriched_terms = FALSE,
  plot_enrichment_chart = FALSE
)

saveRDS(pathfindR_output_Biogrid_Reactome, file=file.path(results_dir, "pathfindR_output_Biogrid_Reactome.RDS"))


pathfindR_output_Biogrid_Reactome_clustered_hierarchical <- cluster_enriched_terms(pathfindR_output_Biogrid_Reactome, plot_dend = FALSE, plot_clusters_graph = FALSE)

saveRDS(pathfindR_output_Biogrid_Reactome_clustered_hierarchical, file=file.path(results_dir, "pathfindR_output_Biogrid_Reactome_clustered_hierarchical.RDS"))

pathfindR_output_Biogrid_Reactome_hierarchical_representative_df <- pathfindR_output_Biogrid_Reactome_clustered_hierarchical[pathfindR_output_Biogrid_Reactome_clustered_hierarchical$Status == "Representative", ] 

pathfindR_output_Biogrid_Reactome_clustered_fuzzy <- cluster_enriched_terms(pathfindR_output_Biogrid_Reactome, plot_dend = FALSE, plot_clusters_graph = FALSE, method = "fuzzy")

pathfindR_output_Biogrid_Reactome_fuzzy_representative_df <- pathfindR_output_Biogrid_Reactome_clustered_fuzzy[pathfindR_output_Biogrid_Reactome_clustered_fuzzy$Status == "Representative", ] 

saveRDS(pathfindR_output_Biogrid_Reactome_clustered_fuzzy, file=file.path(results_dir, "pathfindR_output_Biogrid_Reactome_clustered_fuzzy.RDS"))

# pathfindR_clustered_hierarchical <- cluster_enriched_terms(pathfindR_output_Biogrid_Reactome[1:250, ], plot_dend = FALSE, plot_clusters_graph = FALSE)


pathfindR::term_gene_heatmap(pathfindR_output_Biogrid_Reactome)

enrichment_chart(result_df = pathfindR_output_Biogrid_Reactome, 
                 top_terms = 15)

pr_input_processed <- input_processing(pr_input)

exp_mat <- assay(DeSeq2_return$dds) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"
      ),
      .before = 1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>%
  column_to_rownames("hgnc_symbol") %>% as.matrix()



pathfindR_output_Biogrid_Reactome_enrichment_table <- list(pathfindR_output_Biogrid_Reactome %>% head(10),
         pathfindR_output_Biogrid_Reactome_hierarchical_representative_df %>% head(10))

names(pathfindR_output_Biogrid_Reactome_enrichment_table) <- c("Biogrid Reactome Pathways", "Representative hierarchical clustering")

pathfindR_Reactome_score_matrix <- list()
pathfindR_Reactome_heatmap <- list()
for (i in 1:length(pathfindR_output_Biogrid_Reactome_enrichment_table)) {
  
  pathfindR_Reactome_score_matrix[[i]] <- 
    score_terms(enrichment_table = pathfindR_output_Biogrid_Reactome_enrichment_table[[i]],
                            exp_mat = exp_mat,
                            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            use_description = TRUE, # default FALSE
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"
  

pathfindR_Reactome_heatmap[[i]] <- plot_scores(pathfindR_Reactome_score_matrix[[i]] , 
            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"
 
pathfindR_Reactome_heatmap[[i]] <- pathfindR_Reactome_heatmap[[i]] + scale_y_discrete(labels = function(x) str_wrap(x, 50))
}


pathfindR_Reactome_heatmap[[1]]
ggsave(paste0(plots_dir, Plot_title, "_", Test, "_pathfindR_Reactome_heatmap.png"), height = 3)

pathfindR_Reactome_heatmap[[2]]
ggsave(paste0(plots_dir, Plot_title, "_", Test, "_pathfindR_Reactome_heatmap_clustered.png"), height = 3)
```

```{r}
pathfindR_Reactome_heatmap <- score_terms(enrichment_table = pathfindR_output_Biogrid_Reactome %>% head(30),
                            exp_mat = exp_mat,
                            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            use_description = TRUE, # default FALSE
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"


a <- plot_scores(pathfindR_Reactome_heatmap, 
            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"

a + scale_y_discrete(labels = function(x) str_wrap(x, 50))
ggsave(paste0(plots_dir, Plot_title, "_", Test, "_pathfindR_Reactome_heatmap.png"), height = 8)

```

```{r pathfindR_biocarta}
pathfindR_output_Biogrid_biocarta <- run_pathfindR(
  pr_input,
  gene_sets = "BioCarta",
  pin_name_path = "Biogrid",
  min_gset_size = 5,
  max_gset_size = 300,
  output_dir = file.path(results_dir, "pathfindR_results_biocarta"),
  #iterations=3,
  #visualize_enriched_terms = FALSE,
  plot_enrichment_chart = FALSE
)

pathfindR_output_Biogrid_biocarta_clustered_hierarchical <- cluster_enriched_terms(pathfindR_output_Biogrid_biocarta, plot_dend = FALSE, plot_clusters_graph = FALSE)

pathfindR_output_Biogrid_biocarta_representative_df <- pathfindR_output_Biogrid_biocarta_clustered_hierarchical[pathfindR_output_Biogrid_GO_clustered_hierarchical$Status == "Representative", ]

saveRDS(pathfindR_output_Biogrid_biocarta, file=paste0(results_dir, "pathfindR_output_Biogrid_biocarta.RDS"))


saveRDS(pathfindR_output_Biogrid_biocarta_clustered_hierarchical, file=file.path(results_dir, "pathfindR_output_Biogrid_biocarta_clustered_hierarchical.RDS"))




pathfindR_output_Biogrid_biocarta_enrichment_table <- list(pathfindR_output_Biogrid_biocarta %>% head(30),
         pathfindR_output_Biogrid_biocarta_representative_df %>% head(30))

names(pathfindR_output_Biogrid_biocarta_enrichment_table) <- c("Biogrid biocarta Pathways", "Representative biocarta hierarchical clustering")

pathfindR_biocarta_score_matrix <- list()
pathfindR_biocarta_heatmap <- list()
for (i in 1:length(pathfindR_output_Biogrid_biocarta_enrichment_table)) {
  
  pathfindR_biocarta_score_matrix[[i]] <- 
    score_terms(enrichment_table = pathfindR_output_Biogrid_biocarta_enrichment_table[[i]],
                            exp_mat = exp_mat,
                            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            use_description = TRUE, # default FALSE
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"
  

pathfindR_biocarta_heatmap[[i]] <- plot_scores(pathfindR_biocarta_score_matrix[[i]] , 
            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"
 
pathfindR_biocarta_heatmap[[i]] <- pathfindR_biocarta_heatmap[[i]] + scale_y_discrete(labels = function(x) str_wrap(x, 50))
}


pathfindR_biocarta_heatmap[[1]]
ggsave(paste0(plots_dir, Plot_title, "_", Test, "_pathfindR_biocarta_heatmap.png"), height = 8)

pathfindR_biocarta_heatmap[[2]]
ggsave(paste0(plots_dir, Plot_title, "_", Test, "_pathfindR_biocarta_heatmap_clustered.png"), height = 8)
```


```{r pathfindR_GO}
pathfindR_output_Biogrid_GO <- run_pathfindR(
  pr_input,
  gene_sets = "GO-BP",
  pin_name_path = "Biogrid",
  min_gset_size = 5,
  max_gset_size = 300,
  output_dir = file.path(results_dir, "pathfindR_results_GO"),
  #iterations=3,
  #visualize_enriched_terms = FALSE,
  plot_enrichment_chart = FALSE
)

pathfindR_output_Biogrid_GO_clustered_hierarchical <- cluster_enriched_terms(pathfindR_output_Biogrid_GO, plot_dend = FALSE, plot_clusters_graph = FALSE)

pathfindR_output_Biogrid_GO_representative_df <- pathfindR_output_Biogrid_GO_clustered_hierarchical[pathfindR_output_Biogrid_GO_clustered_hierarchical$Status == "Representative", ]

saveRDS(pathfindR_output_Biogrid_GO, file=paste0(results_dir, "pathfindR_output_Biogrid_GO_BP.RDS"))

enrichment_chart(
  result_df = pathfindR_output_Biogrid_GO_representative_df,
  top_terms = 15
)

pr_input_processed <- input_processing(pr_input)

exp_mat <- assay(DeSeq2_return$dds) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"
      ),
      .before = 1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>%
  column_to_rownames("hgnc_symbol") %>% as.matrix()


pathfindR_GO_score_matrix <- score_terms(enrichment_table = pathfindR_output_Biogrid_GO %>% head(30),
                            exp_mat = exp_mat,
                            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            use_description = TRUE, # default FALSE
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"

pathfindR_GO_heatmap <- plot_scores(pathfindR_GO_score_matrix, 
            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"


pathfindR_GO_heatmap + scale_y_discrete(labels = function(x) str_wrap(x, 50))
ggsave(paste0(plots_dir, Plot_title,"_pathfindR_GO_BP_heatmap.png"), height = 8)



pathfindR_GO_score_matrix_clustered <- score_terms(enrichment_table = pathfindR_output_Biogrid_GO_representative_df %>% head(30),
                            exp_mat = exp_mat,
                            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            use_description = TRUE, # default FALSE
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"

pathfindR_GO_heatmap_clustered <- plot_scores(pathfindR_GO_score_matrix_clustered, 
            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"


pathfindR_GO_heatmap_clustered + scale_y_discrete(labels = function(x) str_wrap(x, 50))
ggsave(paste0(plots_dir, Plot_title,"_pathfindR_GO_BP_heatmap_clustered.png"), height = 8)
```




```{r pathfindR_GO_All}
pathfindR_output_Biogrid_GO_All <- run_pathfindR(
  pr_input,
  gene_sets = "GO-All",
  pin_name_path = "Biogrid",
  min_gset_size = 5,
  max_gset_size = 300,
  output_dir = file.path(results_dir, "pathfindR_results_GO_All"),
  #iterations=3,
  #visualize_enriched_terms = FALSE,
  plot_enrichment_chart = FALSE
)

pathfindR_output_Biogrid_GO_All_clustered_hierarchical <- cluster_enriched_terms(pathfindR_output_Biogrid_GO_All, plot_dend = FALSE, plot_clusters_graph = FALSE)

pathfindR_output_Biogrid_GO_All_representative_df <- pathfindR_output_Biogrid_GO_All_clustered_hierarchical[pathfindR_output_Biogrid_GO_clustered_hierarchical$Status == "Representative", ]

saveRDS(pathfindR_output_Biogrid_GO_All, file=paste0(results_dir, "pathfindR_output_Biogrid_GO_All.RDS"))


saveRDS(pathfindR_output_Biogrid_GO_All_clustered_hierarchical, file=file.path(results_dir, "pathfindR_output_Biogrid_GO_All_clustered_hierarchical.RDS"))

enrichment_chart(
  result_df = pathfindR_output_Biogrid_GO_representative_df,
  top_terms = 15
)

pr_input_processed <- input_processing(pr_input)

exp_mat <- assay(DeSeq2_return$dds) %>%
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
   mutate(
      hgnc_symbol = mapIds(
        EnsDb.Hsapiens.v86,
        keys = id,
        #keytype = "SYMBOL",
        #column = "GENEID"
        keytype = "GENEID",
        column = "SYMBOL"
      ),
      .before = 1
    ) %>%
    drop_na() %>%
    dplyr::select(-id) %>%
  column_to_rownames("hgnc_symbol") %>% as.matrix()


pathfindR_GO_All_score_matrix <- score_terms(enrichment_table = pathfindR_output_Biogrid_GO_All %>% head(30),
                            exp_mat = exp_mat,
                            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            use_description = TRUE, # default FALSE
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"

pathfindR_GO_All_heatmap <- plot_scores(pathfindR_GO_All_score_matrix, 
            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"


pathfindR_GO_All_heatmap + scale_y_discrete(labels = function(x) str_wrap(x, 50))
ggsave(paste0(plots_dir, Plot_title,"_pathfindR_GO_All_heatmap.png"), height = 8)



pathfindR_GO_All_score_matrix_clustered <- score_terms(enrichment_table = pathfindR_output_Biogrid_GO_All_representative_df %>% head(10),
                            exp_mat = exp_mat,
                            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            use_description = TRUE, # default FALSE
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"

pathfindR_GO_All_heatmap_clustered <- plot_scores(pathfindR_GO_All_score_matrix_clustered, 
            cases = colData(DeSeq2_return$dds) %>% as.data.frame() %>% dplyr::filter(.data[[deseq2_design_condition]] == Test) %>% pull(SampleName) %>% as.character(),
                            label_samples = FALSE, # default = TRUE
                            case_title = Test,  # default = "Case"
                            control_title = Control, # default = "Control"
                            low = "#f7797d", # default = "green"
                            mid = "#fffde4", # default = "black"
                            high = "#1f4037") # default = "red"


pathfindR_GO_All_heatmap_clustered + scale_y_discrete(labels = function(x) str_wrap(x, 50))
ggsave(paste0(plots_dir, Plot_title,"_pathfindR_GO_All_heatmap_clustered.png"), height = 3)
```




##DAVID
```{r david_cola}

#BiocManager::install("cola", update=FALSE)
library(cola)

david_enrichment_data <- 
  results_annotated %>% dplyr::filter(padj < 0.05, log2FoldChange > 1) %>%
  dplyr::filter(!is.na(hgnc_symbol) & !duplicated(hgnc_symbol))%>%
  dplyr::pull(ensembl_gene_id)

david_enrichment <- david_enrichment(
    genes = david_enrichment_data,
    email = "anandagopal.srinivasan@ndorms.ox.ac.uk",
    catalog = c("KEGG_PATHWAY"),
    idtype = "ENSEMBL_GENE_ID",
    species = "Homo sapiens")


david_enrichment %>% head(30) %>% kable() %>% kable_classic_2()
```


